src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,188 @@
+               std::numeric_limits<size_t>::max() : maxIterations;
+
+  // initialize the gradients.
+  size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+  arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+  arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+  for (size_t f = 0, b = 0; f < numFunctions;
+    /* incrementing done manually */)
+  {
+    // Find the effective batch size (the last batch may be smaller).
+    effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+    function.Gradient(iterate, f, tableOfGradients.slice(b),
+                      effectiveBatchSize);
+
+    f += effectiveBatchSize;
+    avgGradient += tableOfGradients.slice(b);
+    b++;
+  }
+  avgGradient /= (double) numBatches; // Calculate average gradient
+
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "SAGA: converged to " << overallObjective
+                  << "; terminating  with failure.  Try a smaller step size?"
+                  << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "SAGA: minimized within tolerance " << tolerance
+                  << "; terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        function.Shuffle();
+    }
+
+    // Store current parameter for the calculation of the variance reduced
+    // gradient.
+    iterate0 = iterate;
+
+    batch = math::RandInt(0, numBatches); // Select a random batch
+
+    // Find the effective batch size; we have to take the minimum of three
+    // things:
+    // - the batch size can't be larger than the user-specified batch size;
+    // - the batch size can't be larger than the number of iterations left
+    //       before actualMaxIterations is hit;
+    // - the batch size can't be larger than the number of functions left.
+    effectiveBatchSize = std::min(
+      std::min(batchSize, actualMaxIterations - i),
+      numFunctions - currentFunction);
+
+    batchStart = batch * batchSize;
+
+    // Calculate the gradient of a random function.
+    function.Gradient(iterate, batchStart, gradient,
+                      effectiveBatchSize);
+
+    // Use the update policy to take a step.
+    updatePolicy.Update(iterate, avgGradient, gradient,
+                        tableOfGradients.slice(batch),
+                        stepSize, numBatches);
+    // Update the average gradient.
+    avgGradient += (gradient-tableOfGradients.slice(batch))/numBatches;
```

Comment:
Not sure I get the reason, for removing the inner loop. Maybe I missed something, but dividing by ``numBatches`` in each iteration would result in a really small gradient.

Reply:
Inner loop is removed as per the suggestion of @rcurtin

Reply:
SAGA and SAG counts to the group of variance reduction optimization like SVRG and SARAH. So I think we should follow the SVRG structure instead of SGD (inner loop).

It'll be sum of gradient of all the batches divided by numbatches

I don't think that is implemented, right now you perform a single batch step, and divide the gradient by numBatches:

```c++
avgGradient += (gradient-tableOfGradients.slice(batch))/numBatches;
```

So, I think we should switch back to the inner loop, and incorporate the changes @rcurtin suggested.

Reply:
@zoq Maybe I'm missing something but I think `avgGradient`=(`gradient0` + `gradient1` +....+ `gradientN`)/`numBatches` so suppose `i`th gradient changes where `i` is the random batch selected, then how will `avgGradient` change? We just have to subtract the old `i`th gradient and add the new `i` th gradient i.e. `avgGradient`+= new `i`th gradient/`numBatches` - old `i`th gradient  Let me know if I can clear my point further

Reply:
@rcurtin suggested something like 
```for (iteration = 0, iteration < maxIterations; iteration += batchSize /* this will have to be handled more carefully */)
{
  if (iteration % numPoints == 0)
  {
    // check for convergence, print output
    // also shuffle the dataset if desired
  }

  // select a random batch and update the gradient
}
```
which is the present implementation
my previous implementation was like this

```for (maxIterations)
{
  if (...)
  {
    // print output, check for convergence
  }

  for (batches)
  {
    // select batch randomly, update parameters
  }
}```

So, should i finally switch to previous ones? IMO the present implementaion suggested by @rcurtin  should be correct. Initially I thought it was not suited to SAGA but it works just like the previous implementation just making it clean by incorporating the trick of `actualMaxIteration%numFuntions` which is just like completing one epoch.

Reply:
@rcurtin  can you also take a look here please :) 
Moreover, @zoq  and @rcurtin I think you won't mind but my practical Vivas and Final semester examinations are going on. It'll soon be over and I'll be back on actively contributing. Sorry, for delaying everything :(